### ISSUE TITLE
Sitemap for DocC Hosted Website 

### ISSUE THREAD

**Issue by game-swoosh** [apple/swift-docc#779]
### Feature Name

Sitemap for DocC Hosted Website 

### Description

It seems that DocC does not generate a sitemap. When I navigate to a hosted DocC on a website, I would expect to see sitemap with a list of all the individual links. 

I also tried searching for the sitemap at Apple's Documentation site which is made with DocC at this URL: https://developer.apple.com/documentation/technologies

I can't seem to find it at Apple's Documentation site either. If it does exist, apologies for not seeing it.

Does this exist and if not, is it something that will be added?

**Comment by Jeehut** [comment]
Hey 👋, would like to add to the discussion that a sitemap could help a lot with SEO. In particular, I'm using DocC for the new design of the [wwdcnotes.com](https://wwdcnotes.com) website, but Google doesn't seem to find a sitemap. Not sure if it'll be able to find all notes, the Search Console seems to struggle. So I'd love to see this feature as well. 👍

**Comment by adam-rocska** [comment]
@Jeehut Aren't we yet used to having bad SEO for anything Swift related? I thought it's our style 😅 

_before anyone would take it seriously: it's just a joke I couldn't resist_

**Comment by yo1995** [comment]
I would like to contribute to DocC to support exporting sitemap. I've already got a modified version running locally and successfully exported a sitemap.xml. Just want to clarify a few things:

- Sitemap uses absolute URLs. For our usecase, we need the sitemap when we build the doc with a `--hosting-base-path` argument. As discussed above, DocC doesn't know where the website would be hosted. Does it make sense to only allow generating a sitemap when a hosting base path is provided?
- Would it be good to call this command `generate-sitemap`?

Thanks.

**Comment by d-ronnqvist** [comment]
> * Sitemap uses absolute URLs. For our usecase, we need the sitemap when we build the doc with a `--hosting-base-path` argument. As discussed above, DocC doesn't know where the website would be hosted. Does it make sense to only allow generating a sitemap when a hosting base path is provided?

The `--hosting-base-path` won't do what you need because it only specifies the path prefix, now the absolute URL prefix where the documentation will be hosted. Per the command line documentation:

```
HOSTING OPTIONS:
  --hosting-base-path <hosting-base-path>
                          The base path your documentation website will be
                          hosted at.
        For example, if you deploy your site to
        'example.com/my_name/my_project/documentation' instead of
        'example.com/documentation', pass '/my_name/my_project' as the base
        path.
```

There are other features that would also benefit from knowing the absolute URL where a developer intent to host their documentation. Two such features that come to mind are:
- Adding social sharing metadata (for example [Open Graph](https://ogp.me) where the absolute URL is a required property for every page)
- Establishing links to _external_ documentation dependencies (that are not part of the developer's package documentation).

Adding one (or more) ways to specify an absolute hosting location could be a good topic to pitch in the Swift Forums, with the reason that the knowing absolute hosting location enables other features (at least 3 listed so far).

> * Would it be good to call this command `generate-sitemap`?

To me this mostly depends on what workflow you had in mind. 

If the workflow you're thinking of is that developers first building the documentation and then running a second command to add a sitemap file to the archive output from the first command; then that sounds like a subcommand in the `process-archive` group of commands. If the command processes already-built archives, then it can only work with the JSON data that's available in the archive output.

If you're pitching this command in the forums it might be a good idea to think about how such an archive processing command would work together with combined/merged archives (is it expected that developers generate sitemaps before or after merging?). 

If, on the other hand, the workflow that you're thinking of is that developers build their documentation with a sitemap in a single command; then that sounds more like new (experimental) flag to add to the (main) `convert` command. That would also allow the implementation to access more information than what's available in the output archive.

When if comes to writing pitches for the Swift Forums; [this message](https://forums.swift.org/t/experimental-documentation-coverage/77489/5) briefly summarizes the evolution process for DocC and links to several pitches/proposals—from both people who contribute to DocC frequently and people who are first time contributors—that you can use as inspiration when writing your own pitch/proposal. Alternatively, if you send me a DM on the Swift Forums I can also help you write a short pitch for these changes (or two pitches; if we want to discuss the absolute URL configuration separately from the sitemap command/flag).

**Comment by yo1995** [comment]
I've made the [first pitch](https://forums.swift.org/t/pitch-experimental-absolute-hosting-url/86504) on adding an option for the absolute path. Later on I can create another pitch for the sitemap generation. Thanks for your initial feedback David!

**Comment by adam-rocska** [comment]
Thanks @yo1995 ! It’s promising 👏

**Comment by yo1995** [comment]
Thanks! I'm a bit unclear about the next steps though. I posted the pitch a week ago and haven't received a comment. Should I…
- wait to get some comment
- start working on an implementation and open a PR
- wait for a decision from a member of the DocC workgroup

**Comment by adam-rocska** [comment]
I don’t think you need to wait for a formal “accepted” decision from the DocC workgroup. As far as I understand DocC’s process, pitches are mainly for design feedback and iteration, not a strict Swift Evolution-style gate.

My suggestion would be:

1. Post a short follow-up on the forum pitch saying that you’re planning to start a draft implementation unless there are objections.
2. Open a draft PR for the absolute hosting URL option first, linked to the pitch and this issue.
3. Keep the sitemap generation as a separate follow-up PR/pitch, since it depends on the absolute URL design.

That gives maintainers something concrete to review without forcing them to make an abstract process decision first.

I’m interested in this feature too, and I’d be happy to help with review/testing or implementation work if that’s useful.

**Comment by d-ronnqvist** [comment]
You can definitely start working on this. It's quite common for PRs and pitches to happen in parallel. Just be prepared that forum discussion can move a bit slower and that there might be feedback from the pitch that requires changes to the PR. 

In terms of next steps; a couple suggestions of what you could do next include:
- Open a draft PR for the absolute hosting URL option first. I replied to your pitch today (apologies for the delay) with a few slightly different alternatives. It would be good to get more of the communities inputs on what is the preferred solution but I think that implementation-wise they're all rather similar. I'm only really expecting the spelling of the option(s) to change based on the pitch. Input validation might move from one flag to another but the core logic of it and the specific information that DocC would display if the user provided an absolute URL when only a path-prefix was expected (or vice versa) would remain relevant either way (although might move between the two flags). I think the PR could get to a near-finished state in parallel to the pitch with the public facing spelling(s) of the CLI option being the final thing that's decided in the pitch thread.
- Start thinking about and/or prototyping the sitemap generation. If your primary focus is on _pitching_ a sitemap feature then you mainly have to consider the user-facing impact and workflows; primarily should this be a `--generate-sitemap` flag or a `generate-sitemap` command? If your primary focus is on the _implementation_ of a sitemap feature then you could instead focus on determining what information the code needs in order to create the sitemap and looking to see where that information could come from. If you find that that information is hard to come by in the .doccarchive output, then that could mean that it would be harder to implement this as its own command.
